### ISSUE TITLE
When Using Custom Bone Mapping, Retargeting Animation Glitch/Error

### ISSUE THREAD

**Issue by stryker313** [smix8/GodotAnimationRetargeting#15]
When using custom bone mapping, the retargeted animation either has incorrect naming, or something else is going wrong.
Video: https://youtu.be/9O3UsoxphxE
Steps to reproduce:
  Enter a custom bone mapping for all bones, using two different skeletons with two different naming convenitions
  Click retarget animations
  The resulting animation has the names from the source skeleton, which is an issue because the new skeleton doesn't have those 
  bone names

Also, is there going to be a way to drag a file with a dictionary onto the retarget node so that people can use it without running the game? ( I know you can setup a dictionary in a file, then load using set_custom_bone_mapping())

**Comment by smix8** [comment]
That the old bonetracks with the source bones are kept from copying the animation is a bug.

The retargeting tool is adding bonetracks for all missing bones in the animation but somehow the returned subname for the bonetrack is returned empty. This is why the main node has no name.

I can't see more from the video alone. Can you send me the files?

> Also, is there going to be a way to drag a file with a dictionary onto the retarget node so that people can use it without running the game?

Does Godot Editor Inspector support drag&drop for this? I don't think so, so this needs to wait until I am able to create a custom user interface.

**Comment by stryker313** [comment]
[Animation Retargeter.zip](https://github.com/smix8/GodotAnimationRetargeting/files/6387718/Animation.Retargeter.zip)

**Comment by smix8** [comment]
Thanks

EDIT:
I see what the problem is, should have a fix ready tomorrow.

**Comment by smix8** [comment]
![fixbonemapping](https://user-images.githubusercontent.com/52464204/116332127-718cfa00-a7d1-11eb-8cb1-b7f2be88082f.gif)
@stryker313
Was able to fix mapping and also scale and retarget the animations completely with a position_correction of y-axis -8.62 now. It is good for debugging but for your project I would recommend to use skeletons that are the same scale and have a root bone to avoid precision errors and the manual corrections.

Need to do some code cleanup and, as said, will upload the fix ~~tomorrow~~ in a few hours.

**Comment by smix8** [comment]
I was able to also fix the hip bone position retargeting so manual correction shouldn't be necessary.

Had to add a lot of trackpath comparison to account for different node structures between source and target (something that Godot by default does not care when moving AnimationPlayers or does wrong).

Unfortunatly I am having issues applying all those changes to the module and something broke without an error msg which makes debugging annoying. Will update later.

For now only updated the GDScript 3.x version for testing with commit 093352a3084d097694a782e29a0cab805c9a22eb

**Comment by smix8** [comment]
Fixed by commit c2360db41c74820f207dfc05dd49502aabb01dd9

@stryker313
With the new version I could fix your animation and skeleton 100% with settings:

- root_motion and/or retarget_position = true
- source_skeleton_scale = 10.0

...but as always let me know if something else shows up that isn't working.

EDIT:
Your retarget skeleton is +1 child node deep so all trackpaths were wrong, something that is trouble in default Godot.

I added an auto-rename for trackspaths in this version and it worked for your skeleton to rename all bonetracks from mixamo_xyz to only xyz. Does currently only account for  Transform tracks, not method, property or sound animationtracks.
